### ISSUE TITLE
webkitSpeechRecognition for desktop apps?

### ISSUE THREAD

**Issue by isimmons** [nwjs/nw.js#1115]
I was attempting to add speech recognition to a desktop application but it seems speechRecognition is not working probably due to the application using local file urls and no way to allow permission like when running it in chrome.

https://github.com/TalAter/annyang/issues/44

As you can see in the linked related issue trying to run speecheRecognition by it's self in the console does nothing in node-webkit.

```
var sr = new webkitSpeechRecognition;
sr.continuous = true; 
sr.interimResults = true;
sr.lang='en';
sr.onresult = function(e) {
    console.log(e.results[e.results.length-1][0].transcript);
};
sr.start();//nothin happens and no errors
```

So would it be correct to say this is not currently possible in node-webkit or is there a known work around? I'm aware of the getUserMedia API but it seems that only captures audio but doesn't do any speech recognition. 

If it's not currently possible will it be possible in future releases?
##

**Comment by mpreziuso** [comment]
Please see #2051 :)

**Comment by JaredCubilla** [comment]
+1

**Comment by rogerwang** [comment]
The feature seems useless because in order to use this you'll need an API key, under which google limited with 50 requests per day ...

https://groups.google.com/a/chromium.org/d/msg/chromium-dev/6HDtCUfIbLE/3GExi4-XnMcJ

**Comment by jhermsmeier** [comment]
Maybe fairly useless with Google's API, but there's a `.serviceURI` property on a `SpeechRecognition` instance: https://dvcs.w3.org/hg/speech-api/raw-file/tip/speechapi.html#api_description
which enables use with other service endpoints (i.e. even a machine local one, which is also what I was intending to experiment with):

**Comment by miller9904** [comment]
Good point.  I would love to be able to plug in my own speech recognizer.

**Comment by rogerwang** [comment]
@jhermsmeier good to know it! do you happen to know any service with which I can test my commit ?

**Comment by jhermsmeier** [comment]
@rogerwang Found WAMI from MIT's CSAIL, it even supports several languages (tested English & German so far): `.serviceURI = 'wami.csail.mit.edu'`

**Comment by RobinMalfait** [comment]
Jonas, could you make an example file how to use that? thanks


Met vriendelijke groeten
Robin Malfait 



Web Developer | Web Designer 




Claever Associates 

Weertstraat 66
2880 Bornem 

+32 (0)3 296 59 17
  

www.claever.be

**Comment by jhermsmeier** [comment]
@RobinMalfait here's a quick thing I threw together: http://bl.ocks.org/jhermsmeier/3bc995d37f3acc0b0364

**Comment by willemmulder** [comment]
That works great! :-)

**Comment by Gimmeaphatbeat** [comment]
@willemmulder from Roger's comments it looks like he's working on adding support for it.

**Comment by willemmulder** [comment]
Ah thanks, I missed the commit a few messages up. Excellent, thanks @rogerwang

**Comment by jhermsmeier** [comment]
@rogerwang is there any progress on this? (not pushing, just curious what the current state of things is)

**Comment by rogerwang** [comment]
@jhermsmeier I plan to add it in one of the beta versions of 0.12.0.

**Comment by rogerwang** [comment]
unfortunately there is no support for 'serviceURI' in current implementation of upstream: https://chromium.googlesource.com/chromium/blink/+/master/Source/modules/speech/SpeechRecognition.idl
so we may have to wait for upstream for it.

The current git version of NW supports google service but lacks usable API key...

**Comment by jhermsmeier** [comment]
Nice, thanks for the update! Wondering why `serviceURI` isn't in upstream yet, though. It's been in stable Chrome for a while.

**Comment by willemmulder** [comment]
Thanks for the update! I hope ServiceURI becomes available soon, then we can really start doing nice stuff!

**Comment by barbass** [comment]
+1

**Comment by kalaomer** [comment]
+1

**Comment by emroot** [comment]
+1

**Comment by sholladay** [comment]
:+1:

**Comment by motoko-11** [comment]
+1

**Comment by jaybrownlee** [comment]
+1

**Comment by vasametropolis** [comment]
+1

**Comment by jhm-ciberman** [comment]
+1

**Comment by KingRial** [comment]
+1

**Comment by TemaSM** [comment]
+1

**Comment by zackm0571** [comment]
+1

**Comment by tom-s** [comment]
+1

**Comment by Aaronik** [comment]
+1

**Comment by ghostoy** [comment]
It should be supported by NW13.

Aaron Sullivan <notifications@github.com>于2015年12月25日周五 00:58写道：

**Comment by jhm-ciberman** [comment]
Is this supported already? I'm using NW13 beta 2. 
And this code throw a SpeechRecognitionError (error="network", message="")

var recognition = new webkitSpeechRecognition();
recognition.onresult = function(e) { console.log(e);}
recognition.onerror = function(e) {console.log(e);}
// Fails with and without this line: 
//recognition.serviceURI = 'wami.csail.mit.edu';
recognition.start();

**Comment by willemmulder** [comment]
I assume wami.csail.mit.edu returns an error or requires a key of sorts? What do you see in the network requests?

**Comment by JonasT** [comment]
Does this feature work without embedding black box binary blob from Google that accesses the microphone without user consent (and without any web app making explicit use of it)?

Please note while I might sound paranoid, this is apparently what chromium used to do to support this: https://www.privateinternetaccess.com/blog/2015/06/google-chrome-listening-in-to-your-room-shows-the-importance-of-privacy-defense-in-depth/

If this is what you plan to add to nwjs as well, maybe make the binary blob download only triggered at runtime by the web app so no app using nwjs is FORCED to run this potential spyware code, or if this needs to be embedded into nwjs from now on then please provide a separate download that doesn't contain this feature.

TL;DR: make sure apps that don't want to use speech recognition don't need to get shipped with nw.js versions that potentially spy on users with the microphone without asking. Again, see the URL above.

**Comment by jhm-ciberman** [comment]
If I use the code above in NW, it fails with a SpeechRecognitionError. If I try (with or without the wami.csail.mit.edu) in regular Google Chrome, it works fine and detects the speech correctly. 
In neither case network requests are generated or showed in the devtools panel.

**Comment by Nelderson** [comment]
+1

**Comment by frenchbread** [comment]
+1

**Comment by jhermsmeier** [comment]
@jhm-ciberman @willemmulder I just recently ran into the same network error with the `0.14.0-sdk`, might get around to looking into what's going on there in the next days.

**Comment by willemmulder** [comment]
Thanks @jhermsmeier ! Really looking forward to your findings.

**Comment by jhermsmeier** [comment]
I've just done some testing with Chrome, Chromium and nw.js - attempting to use a local speech recognition backend running on `localhost` via `serviceURI`. First finding: The `SpeechRecognition` API stopped giving a flying fuck about the `serviceURI` property. It's completely ignored from what I can tell... i.e. with Chromium no connection to `localhost` could be observed, yet the speech recognition worked perfectly fine. Might find the time to dig into the source and file a bug, if there isn't one already.

**Comment by willemmulder** [comment]
Hmm I'm pretty sure that using a different URL (I used wami.csail.mit.edu) worked before in Chrome. However, if I read this thread https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/82LcTDrhshw then there's some talks about not supporting a full URL, whatever that means. Or in other words: I'm just as surprised as you are.

**Comment by jhermsmeier** [comment]
Looks like it's been removed / reverted 4 months ago: https://bugs.chromium.org/p/chromium/issues/detail?id=480516#c6

```diff
Index: third_party/WebKit/Source/modules/speech/SpeechRecognition.idl
diff --git a/third_party/WebKit/Source/modules/speech/SpeechRecognition.idl b/third_party/WebKit/Source/modules/speech/SpeechRecognition.idl
index 8b3b25b90a637627bb02ddb19aa2c1c147293be0..4e13bbb7d7b51dbf76b84cea5bdf672767ee13a7 100644
--- a/third_party/WebKit/Source/modules/speech/SpeechRecognition.idl
+++ b/third_party/WebKit/Source/modules/speech/SpeechRecognition.idl
@@ -32,7 +32,6 @@
 ] interface SpeechRecognition : EventTarget {
     [LegacyInterfaceTypeChecking] attribute SpeechGrammarList grammars;
     attribute DOMString lang;
-    attribute DOMString serviceURI;
     attribute boolean continuous;
     attribute boolean interimResults;
     attribute unsigned long maxAlternatives;
```

Can't see a valid reason for why they did though, as from what I can see, it's still in the draft/spec (although on the other hand; that's in flux, and I have difficulties finding a recent document).

There are some mentions of going through `chrome://speech-recognition` to use custom recognizers – which I haven't quite fully understood, yet. Will really need to dig into the issue tracker and source to figure out what's happening with that I guess.

**Comment by bromagosa** [comment]
Any updates on this?

`chrome.tts` doesn't seem to work at all in nwjs.io, neither does `window.speechSynthesis`. Has anyone succeeded?

I thought it may be an issue of adding the right permissions to `ttsEngine` in the manifest, but that doesn't seem to do the trick either.

**Comment by rogerwang** [comment]
@bromagosa I just tested chrome.tts works for me. This issue is about speech recognition, not text-to-speech. Please file another issue if it has bug on your side.

**Comment by rogerwang** [comment]
To all: I just had a test, speech recognition is working with a valid google API key. So I'm closing this issue. 

I'm testing with https://www.google.com/intl/en/chrome/demos/speech.html.

See here to see how to get a key: https://www.chromium.org/developers/how-tos/api-keys

**Comment by bromagosa** [comment]
@rogerwang sorry, you're right. I was hunting for solutions for both because speech recognition wasn't working either, but I'll take a look at your links. Thanks!

p.s. I found out `chrome.tts` is supported in Windows, Mac and Chromebooks, but (seemingly arbitrarily) not in GNU/Linux, so that's why I can't get it to work. I may need to work around that by adding a Node binding to festival :(

**Comment by mscreenie** [comment]
The Speech API looks great but the current limits imposed by Google make it unusable for developers. 

There needs to be a really good offline ASR so not to ever worry about Google and being online but that probably wont happen anytime soon.

**Comment by Sugarcaen** [comment]
It looks like you could use the Google Cloud Platform for speech rec, or spawn a chrome process and try piping everything through there.

**Comment by mscreenie** [comment]
Interesting. Could you elaborate?

**Comment by mscreenie** [comment]
I think @jhermsmeier comment is interesting. **serviceURI** was removed in Chromium 49 and I can't find any information as to why. I know the API and spec is draft. 

This really sucks because vendor lock in and to implement your own self hosted speech to text is now really difficult. Frustrating to say the least - Don't know if this is Google wanting all the info to flow into their veins.

**Comment by jhermsmeier** [comment]
According to the [latest draft](https://dvcs.w3.org/hg/speech-api/raw-file/tip/webspeechapi.html#speechreco-section) of the Web Speech API, the `serviceURI` hasn't been removed from the specification.

So I went digging a bit, which did turn up some things I haven't discovered before, but most discussion around this seems to have happened out of band, or somewhere inaccessible to the public.

There was [Chromium Issue N° 480516](https://bugs.chromium.org/p/chromium/issues/detail?id=480516) filed ~2 years ago, in which support was added, and subsequently reverted again a few months later. Discussion related to this happened on [blink-dev/82LcTDrhshw/pGKPgrXOUaAJ](https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/82LcTDrhshw/pGKPgrXOUaAJ), yet it is still unclear to me what exactly happened there.

It might be worth filing a new Chromium Issue to get an update on the situation.

**Comment by mscreenie** [comment]
Thanks for digging. I've done some myself and have come to the same conclusion. Everything is still very unclear and seems odd `serviceURI` to be removed and I can't establish why. It's a little frustrating.

I am hopeful it returns but for now I'm exploring other things. Firefox launched VoiceFill which has similar functionality, I haven't dug into how it works completely but this is helpful:

https://github.com/mozilla/speech-proxy

VoiceFill is also written as a WebExtension unless I am mistaken. If only Chromium had WebExtension support there could be hope to port it as a viable replacement.

**Comment by vsharmaMitel** [comment]
Hello, may I know why this issue was closed. I really want to implement webkitSpeechRecognition for my nodewebkit app. Can someone please tell me if there is any solution or workaround?

**Comment by ignStonechat** [comment]
+1

**Comment by ignStonechat** [comment]
I'm searching for Electron alternative for my app just for speech recognition and I hope this works!

**Comment by mscreenie** [comment]
Chromium has recently release SODA (Works offline). This may be a better solution than this.
